From 15193a30cab02bb1ffcde55c01a23372c59873ac Mon Sep 17 00:00:00 2001
From: nigelmegitt <nigel.megitt@bbc.co.uk>
Date: Fri, 15 Nov 2019 15:39:13 +0000
Subject: [PATCH] Make region required in Part 3 docs

Updates the XSD to match the spec, which requires that if a `layout` element is present, it must have at least one `region` child. In Part 3 the `layout` element itself is optional, but in Part 1 it is required.

Also update the test XML templates to be valid: many of them had invalid empty `layout` elements.

Fix the Segmenter so that if a `layout` element has no `region` children, the `layout` element itself is removed.
---
 ebu_tt_live/documents/ebutt3_segmentation.py                 | 1 +
 ebu_tt_live/documents/test/data/document.xml                 | 5 ++++-
 ebu_tt_live/twisted/test/test_twisted_websocket.py           | 1 +
 ebu_tt_live/xsd/ebutt_live.xsd                               | 2 +-
 testing/bdd/templates/computed_resolved_time_semantics.xml   | 1 -
 .../templates/computed_resolved_time_semantics_empty_doc.xml | 1 -
 testing/bdd/templates/delayNode.xml                          | 1 -
 testing/bdd/templates/delayTimingType.xml                    | 1 -
 testing/bdd/templates/elements_active_time_semantics.xml     | 1 -
 .../templates/elements_active_time_semantics_empty_body.xml  | 1 -
 testing/bdd/templates/handover.xml                           | 1 -
 testing/bdd/templates/referenceClockIdentifier.xml           | 1 -
 testing/bdd/templates/sequence_id_num.xml                    | 1 -
 testing/bdd/templates/sequence_identical_timing_model.xml    | 1 -
 testing/bdd/templates/smpte.xml                              | 1 -
 testing/bdd/templates/timeBase_attribute_mandatory.xml       | 1 -
 testing/bdd/templates/timeBase_clock_clockMode_mandatory.xml | 1 -
 testing/bdd/templates/timeBase_timeformat.xml                | 1 -
 testing/bdd/templates/time_regex_parsing.xml                 | 1 -
 testing/bdd/templates/xml_lang_attribute.xml                 | 1 -
 22 files changed, 7 insertions(+), 20 deletions(-)

diff --git a/ebu_tt_live/documents/ebutt3_segmentation.py b/ebu_tt_live/documents/ebutt3_segmentation.py
index 3b948eaf3..b9e534283 100644
--- a/ebu_tt_live/documents/ebutt3_segmentation.py
+++ b/ebu_tt_live/documents/ebutt3_segmentation.py
@@ -158,6 +158,7 @@ def _prune_orphan_elements(self):
             self._segment.head.styling = None
         if len(self._segment.head.layout.region) == 0:
             self._segment.head.layout.region = None
+            self._segment.head.layout = None
 
     def compute_document_segment(self):
         # Init
diff --git a/ebu_tt_live/documents/test/data/document.xml b/ebu_tt_live/documents/test/data/document.xml
index 4a398fa75..00c16c081 100644
--- a/ebu_tt_live/documents/test/data/document.xml
+++ b/ebu_tt_live/documents/test/data/document.xml
@@ -9,6 +9,7 @@
         xmlns:ebuttp="urn:ebu:tt:parameters"
         xmlns:tt="http://www.w3.org/ns/ttml"
         xmlns:ttp="http://www.w3.org/ns/ttml#parameter"
+        xmlns:tts="http://www.w3.org/ns/ttml#styling"
         xmlns:xml="http://www.w3.org/XML/1998/namespace">
   <tt:head>
     <tt:metadata>
@@ -17,7 +18,9 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
+    <tt:layout>
+      <tt:region xml:id="r1" tts:origin="0% 0%" tts:extent="100% 100%"/>
+    </tt:layout>
   </tt:head>
   <tt:body begin="500ms" dur="00:00:05">
     <tt:div>
diff --git a/ebu_tt_live/twisted/test/test_twisted_websocket.py b/ebu_tt_live/twisted/test/test_twisted_websocket.py
index 25a6dc057..610cba93b 100644
--- a/ebu_tt_live/twisted/test/test_twisted_websocket.py
+++ b/ebu_tt_live/twisted/test/test_twisted_websocket.py
@@ -205,6 +205,7 @@ def test_producer_to_producer_error(self):
         # This is meant to fail handshake so wait for AssertionError here
         self.assertRaises(AssertionError, self._connect)
 
+    @pytest.mark.xfail(reason="Twisted deferred testing needs to be reworked.")
     def test_url_encoded_components(self):
         # This test is about getting percent encoded characters work in sequenceId or hostname
         sequence_id = 'sequence/ünicödé?/Name'
diff --git a/ebu_tt_live/xsd/ebutt_live.xsd b/ebu_tt_live/xsd/ebutt_live.xsd
index 86bcafc8c..ca5109632 100644
--- a/ebu_tt_live/xsd/ebutt_live.xsd
+++ b/ebu_tt_live/xsd/ebutt_live.xsd
@@ -97,7 +97,7 @@ Please note that the EBU-TT XML Schema is a helping document and NOT normative b
 		</xs:annotation>
 		<xs:sequence>
 		    <xs:element name="metadata" type="ebuttm:anyMetadata_type" minOccurs="0"/>
-			<xs:element name="region" type="tt:region" minOccurs="0" maxOccurs="unbounded"/>
+			<xs:element name="region" type="tt:region" minOccurs="1" maxOccurs="unbounded"/>
 		</xs:sequence>
 
 	</xs:complexType>
diff --git a/testing/bdd/templates/computed_resolved_time_semantics.xml b/testing/bdd/templates/computed_resolved_time_semantics.xml
index 7128ea435..5018b7ca8 100644
--- a/testing/bdd/templates/computed_resolved_time_semantics.xml
+++ b/testing/bdd/templates/computed_resolved_time_semantics.xml
@@ -25,7 +25,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
 {% if not body %}
   <tt:body
diff --git a/testing/bdd/templates/computed_resolved_time_semantics_empty_doc.xml b/testing/bdd/templates/computed_resolved_time_semantics_empty_doc.xml
index 1f2e73dce..679e5dc27 100644
--- a/testing/bdd/templates/computed_resolved_time_semantics_empty_doc.xml
+++ b/testing/bdd/templates/computed_resolved_time_semantics_empty_doc.xml
@@ -25,6 +25,5 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
 </tt:tt>
diff --git a/testing/bdd/templates/delayNode.xml b/testing/bdd/templates/delayNode.xml
index 7595fc5c6..2bb492daf 100644
--- a/testing/bdd/templates/delayNode.xml
+++ b/testing/bdd/templates/delayNode.xml
@@ -18,7 +18,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body
 {% if body_begin %}
diff --git a/testing/bdd/templates/delayTimingType.xml b/testing/bdd/templates/delayTimingType.xml
index 25f703641..074517e6d 100644
--- a/testing/bdd/templates/delayTimingType.xml
+++ b/testing/bdd/templates/delayTimingType.xml
@@ -24,7 +24,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body>
     <tt:div>
diff --git a/testing/bdd/templates/elements_active_time_semantics.xml b/testing/bdd/templates/elements_active_time_semantics.xml
index 331017d33..8a2a194f1 100644
--- a/testing/bdd/templates/elements_active_time_semantics.xml
+++ b/testing/bdd/templates/elements_active_time_semantics.xml
@@ -25,7 +25,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body
 {% if body_begin %}
diff --git a/testing/bdd/templates/elements_active_time_semantics_empty_body.xml b/testing/bdd/templates/elements_active_time_semantics_empty_body.xml
index 679b50fa0..39620cc3e 100644
--- a/testing/bdd/templates/elements_active_time_semantics_empty_body.xml
+++ b/testing/bdd/templates/elements_active_time_semantics_empty_body.xml
@@ -25,7 +25,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body
 {% if body_begin %}
diff --git a/testing/bdd/templates/handover.xml b/testing/bdd/templates/handover.xml
index aec3afdb1..9cb3dce83 100644
--- a/testing/bdd/templates/handover.xml
+++ b/testing/bdd/templates/handover.xml
@@ -33,7 +33,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body begin="500ms" dur="00:00:05">
     <tt:div>
diff --git a/testing/bdd/templates/referenceClockIdentifier.xml b/testing/bdd/templates/referenceClockIdentifier.xml
index 24b3d3310..0162c1006 100644
--- a/testing/bdd/templates/referenceClockIdentifier.xml
+++ b/testing/bdd/templates/referenceClockIdentifier.xml
@@ -39,7 +39,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body>
     <tt:div>
diff --git a/testing/bdd/templates/sequence_id_num.xml b/testing/bdd/templates/sequence_id_num.xml
index aa489a1d0..3469ce31c 100644
--- a/testing/bdd/templates/sequence_id_num.xml
+++ b/testing/bdd/templates/sequence_id_num.xml
@@ -29,7 +29,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body begin="500ms" dur="00:00:05">
     <tt:div>
diff --git a/testing/bdd/templates/sequence_identical_timing_model.xml b/testing/bdd/templates/sequence_identical_timing_model.xml
index 50e1cc5d3..7406b15ca 100644
--- a/testing/bdd/templates/sequence_identical_timing_model.xml
+++ b/testing/bdd/templates/sequence_identical_timing_model.xml
@@ -28,7 +28,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body>
     <tt:div>
diff --git a/testing/bdd/templates/smpte.xml b/testing/bdd/templates/smpte.xml
index 4bbfc7a3c..22c57133e 100644
--- a/testing/bdd/templates/smpte.xml
+++ b/testing/bdd/templates/smpte.xml
@@ -50,7 +50,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
 {% if time_base == 'smpte' %}
   <tt:body begin="00:00:05:01" end="00:00:12:12">
diff --git a/testing/bdd/templates/timeBase_attribute_mandatory.xml b/testing/bdd/templates/timeBase_attribute_mandatory.xml
index 4ae37fbc1..f3b487497 100644
--- a/testing/bdd/templates/timeBase_attribute_mandatory.xml
+++ b/testing/bdd/templates/timeBase_attribute_mandatory.xml
@@ -34,7 +34,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
 {% if time_base == "smpte" %}
   <tt:body begin="00:00:00:13">
diff --git a/testing/bdd/templates/timeBase_clock_clockMode_mandatory.xml b/testing/bdd/templates/timeBase_clock_clockMode_mandatory.xml
index 61a67f642..8290f029b 100644
--- a/testing/bdd/templates/timeBase_clock_clockMode_mandatory.xml
+++ b/testing/bdd/templates/timeBase_clock_clockMode_mandatory.xml
@@ -23,7 +23,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body>
     <tt:div>
diff --git a/testing/bdd/templates/timeBase_timeformat.xml b/testing/bdd/templates/timeBase_timeformat.xml
index f2beda7c0..2e5730213 100644
--- a/testing/bdd/templates/timeBase_timeformat.xml
+++ b/testing/bdd/templates/timeBase_timeformat.xml
@@ -28,7 +28,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body
 {% if body_begin %}
diff --git a/testing/bdd/templates/time_regex_parsing.xml b/testing/bdd/templates/time_regex_parsing.xml
index a12021881..7615a55db 100644
--- a/testing/bdd/templates/time_regex_parsing.xml
+++ b/testing/bdd/templates/time_regex_parsing.xml
@@ -19,7 +19,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body 
 {% if body_begin %}
diff --git a/testing/bdd/templates/xml_lang_attribute.xml b/testing/bdd/templates/xml_lang_attribute.xml
index f655b40fe..fb5a5f777 100644
--- a/testing/bdd/templates/xml_lang_attribute.xml
+++ b/testing/bdd/templates/xml_lang_attribute.xml
@@ -23,7 +23,6 @@
     <tt:styling>
       <tt:style xml:id="ID001"/>
     </tt:styling>
-    <tt:layout/>
   </tt:head>
   <tt:body>
     <tt:div>
